index/segment/fs/segment.go

```diff
@@ -0,0 +1,477 @@
+	defer fstCloser.Close()
+
+	postingsOffset, exists, err := termsFST.Get(term)
+	if err != nil {
+		return nil, err
+	}
+
+	if !exists {
+		// i.e. we don't know anything about the term, so can early return an empty postings list
+		return r.opts.PostingsListPool.Get(), nil
+	}
+
+	pl, err := r.retrievePostingsListWithRLock(postingsOffset)
+	if err != nil {
+		return nil, err
+	}
+
+	if err := fstCloser.Close(); err != nil {
+		return nil, err
+	}
+
+	return pl, nil
+}
+
+func (r *segment) MatchRegexp(field []byte, regexp []byte, compiled *regexp.Regexp) (postings.List, error) {
+	r.RLock()
+	defer r.RUnlock()
+	if r.closed {
+		return nil, errReaderClosed
+	}
+
+	re, err := vregex.New(string(regexp))
+	if err != nil {
+		return nil, err
+	}
+
+	termsFST, err := r.retrieveTermsFSTWithRLock(field)
+	if err != nil {
+		return nil, err
+	}
+
+	var (
+		fstCloser     = x.NewSafeCloser(termsFST)
+		pl            = r.opts.PostingsListPool.Get()
+		iter, iterErr = termsFST.Search(re, minByteKey, maxByteKey)
+		iterCloser    = x.NewSafeCloser(iter)
+	)
+	defer func() {
+		iterCloser.Close()
+		fstCloser.Close()
+	}()
+
+	for {
+		if iterErr == vellum.ErrIteratorDone {
+			break
+		}
+
+		if iterErr != nil {
+			return nil, iterErr
+		}
+
+		_, postingsOffset := iter.Current()
+		nextPl, err := r.retrievePostingsListWithRLock(postingsOffset)
+		if err != nil {
+			return nil, err
+		}
+		if err := pl.Union(nextPl); err != nil {
+			return nil, err
+		}
+
+		iterErr = iter.Next()
+	}
+
+	if err := iterCloser.Close(); err != nil {
+		return nil, err
+	}
+
+	if err := fstCloser.Close(); err != nil {
+		return nil, err
+	}
+
+	return pl, nil
+}
+
+func (r *segment) MatchAll() (postings.MutableList, error) {
+	return nil, errNotImplemented
+}
+
+func (r *segment) Docs(pl postings.List) (doc.Iterator, error) {
+	return nil, errNotImplemented
+}
+
+func (r *segment) AllDocs() (doc.Iterator, error) {
+	return nil, errNotImplemented
+}
+
+func (r *segment) retrievePostingsListWithRLock(postingsOffset uint64) (postings.List, error) {
+	postingsBytes, err := r.retrieveBytesWithRLock(r.opts.PostingsData, postingsOffset)
+	if err != nil {
+		return nil, fmt.Errorf("unable to retrieve postings data: %v", err)
+	}
+
+	return pilosa.Unmarshal(postingsBytes, roaring.NewPostingsList)
+}
+
+func (r *segment) allKeys(fst *vellum.FST) ([][]byte, error) {
+	num := fst.Len()
+	keys := make([][]byte, 0, num)
+
+	iter, iterErr := fst.Iterator(minByteKey, maxByteKey)
+	safeCloser := x.NewSafeCloser(iter)
+	defer safeCloser.Close()
+
+	for {
+		if iterErr == vellum.ErrIteratorDone {
+			break
+		}
+		if iterErr != nil {
+			return nil, iterErr
+		}
+		key, _ := iter.Current()
+		keys = append(keys, r.copyBytes(key))
+		iterErr = iter.Next()
+	}
+
+	if err := safeCloser.Close(); err != nil {
+		return nil, err
+	}
+	return keys, nil
+}
+
+func (r *segment) retrieveTermsFSTWithRLock(field []byte) (*vellum.FST, error) {
+	termsFSTOffset, exists, err := r.fieldsFST.Get(field)
```

Comment:
Would it be worth caching the terms FSTs in a map from offset to FST so we don't have to reload them each time?

Reply:
Yes, can be a subsequent PR based on perf work we have do to for query anyway?

Reply:
It'd have to be file backed, we can't keep it in mem or if we do ideally it'd be a mmap'd backed map (which is actually kinda frustrating to implement).